### PR TITLE
SO-4798: Migrate module restricted rules

### PR DESCRIPTION
**Pull request by molnarlaura**

### CODE REVIEW

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/validation/SnomedQueryValidationRuleEvaluator.java [hunk 1]
```diff
@@ -143,6 +130,7 @@ public String type() {
 		@JsonProperty private Boolean active;
 		@JsonProperty private String effectiveTime;
 		@JsonProperty private String module;
+		@JsonProperty private Boolean filterByWorkingModule;
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/validation/SnomedQueryValidationRuleEvaluator.java]: It is possible to define a default value for a JsonProperty (by using `defaultValue`). I would suggest setting this value to false by default, so then you don't need to change so many rule declarations. This way we will have three ideal scenarios (and a not-quite-ideal fourth):
- there is neither `module` nor `filterByWorkingModule` value specified in the rule implementation --> do not apply any module filtering
- there is no `module` value but `filterByWorkingModule = true` --> apply module filtering based on the default code system modules
- there is a `module` value (e.g. an ECL expression) and `filterByWorkingModule` is not present (== false) --> filter by module using the expression in the property `module`
- there is a `module` value and `filterByWorkingModule = true` --> this is a contradiction but we can still say to use the value in the `module` parameter and filter by that, assuming that must be something more specific provided on purpose

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/validation/SnomedQueryValidationRuleEvaluator.java [hunk 2]
```diff
@@ -155,11 +143,14 @@ public final SB prepareSearch() {
...
 		protected SB prepareSearch(SB req) {
-			return req.filterByActive(active)
-					.filterByReleased(released)
-					.filterByModule(module)
-					.filterByEffectiveTime(effectiveTime);
+			req.filterByActive(active)
+			.filterByReleased(released)
+			.filterByEffectiveTime(effectiveTime);
 			
+			if (!filterByWorkingModule) {
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/validation/SnomedQueryValidationRuleEvaluator.java]: Change this condition according to the cases mentioned above.

File: snomed/com.b2international.snowowl.validation.snomed/src/main/resources/scripts/rule532a.groovy [hunk 3]
```diff
@@ -0,0 +1,88 @@
+package scripts
+
+import static com.b2international.snowowl.snomed.common.SnomedConstants.Concepts.UK_MAINTAINED_CLINICAL_MODULE
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.validation.snomed/src/main/resources/scripts/rule532a.groovy]: Please make sure no UK content leaks through to this repo. Pro tip: Ctrl + Shift + O will organize the import statements in any Groovy file

File: snomed/com.b2international.snowowl.validation.snomed/src/main/resources/scripts/rule532b.groovy [hunk 4]
```diff
@@ -0,0 +1,98 @@
+package scripts
+
+import static com.b2international.snowowl.snomed.common.SnomedConstants.Concepts.UK_MAINTAINED_CLINICAL_MODULE
+import static com.b2international.snowowl.snomed.common.SnomedConstants.Concepts.UK_MAINTAINED_PHARMACY_MODULE
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.validation.snomed/src/main/resources/scripts/rule532b.groovy]: Same here

File: snomed/com.b2international.snowowl.validation.snomed/src/main/resources/scripts/rule482.groovy [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+import com.b2international.snowowl.snomed.datastore.index.entry.SnomedDocument
+import com.b2international.snowowl.snomed.datastore.request.SnomedRequests
+import com.google.common.base.Strings
+import com.google.common.collect.ImmutableMap
+
+import groovy.transform.Field
+
+List<ComponentIdentifier> issues = new ArrayList<>()
+
+def ukModules = SnomedRequests.prepareSearchConcept()
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.validation.snomed/src/main/resources/scripts/rule482.groovy]: Please rename this variable

File: snomed/com.b2international.snowowl.validation.snomed/src/test/java/com/b2international/snowowl/validation/snomed/BaseGenericValidationRuleTest.java [hunk 6]
```diff
@@ -92,7 +94,8 @@ protected void configureContext(Builder context) {
 	@Override
 	protected void configureValidationRequest(ValidateRequestBuilder req) {
 		req.setRuleParameters(Map.of(
-			ValidationConfiguration.IS_UNPUBLISHED_ONLY, effectiveTime == EffectiveTimes.UNSET_EFFECTIVE_TIME
+			ValidationConfiguration.IS_UNPUBLISHED_ONLY, effectiveTime == EffectiveTimes.UNSET_EFFECTIVE_TIME,
+			ValidationConfiguration.WORKING_MODULES, "<<" + Concepts.UK_DRUG_EXTENSION_MODULE
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.validation.snomed/src/test/java/com/b2international/snowowl/validation/snomed/BaseGenericValidationRuleTest.java]: You can also use a customer independent module ID for the unit tests. We have our own module called `MODULE_B2I_EXTENSION`

**Review by nagyo — CHANGES_REQUESTED** [review]

File: snomed/com.b2international.snowowl.validation.snomed/src/main/resources/scripts/rule532a.groovy (hunk 3 above)
Comment (molnarlaura) [reply snomed/com.b2international.snowowl.validation.snomed/src/main/resources/scripts/rule532a.groovy]: Done

File: snomed/com.b2international.snowowl.validation.snomed/src/main/resources/scripts/rule532b.groovy (hunk 4 above)
Comment (molnarlaura) [reply snomed/com.b2international.snowowl.validation.snomed/src/main/resources/scripts/rule532b.groovy]: Done

File: snomed/com.b2international.snowowl.validation.snomed/src/main/resources/scripts/rule482.groovy (hunk 5 above)
Comment (molnarlaura) [reply snomed/com.b2international.snowowl.validation.snomed/src/main/resources/scripts/rule482.groovy]: Done

File: snomed/com.b2international.snowowl.validation.snomed/src/test/java/com/b2international/snowowl/validation/snomed/BaseGenericValidationRuleTest.java (hunk 6 above)
Comment (molnarlaura) [reply snomed/com.b2international.snowowl.validation.snomed/src/test/java/com/b2international/snowowl/validation/snomed/BaseGenericValidationRuleTest.java]: Done

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/validation/SnomedQueryValidationRuleEvaluator.java (hunk 2 above)
Comment (molnarlaura) [reply snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/validation/SnomedQueryValidationRuleEvaluator.java]: Will move the logic to `eval()`.

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/validation/SnomedQueryValidationRuleEvaluator.java (hunk 1 above)
Comment (molnarlaura) [reply snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/validation/SnomedQueryValidationRuleEvaluator.java]: Done

File: snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/validation/SnomedQueryValidationRuleEvaluator.java [hunk 7]
```diff
@@ -89,6 +72,16 @@
 		SnomedSearchRequestBuilder<?, PageableCollectionResource<SnomedComponent>> req = validationQuery
 				.prepareSearch();
 		
+		String extensionModules = params != null && params.containsKey(ValidationConfiguration.MODULES) ? params.get(ValidationConfiguration.MODULES).toString() : "";
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/validation/SnomedQueryValidationRuleEvaluator.java]: That toString() could cause an NPE, feel free to cast the value to String as is. E.g.:
```suggestion
		String extensionModules = params != null && params.containsKey(ValidationConfiguration.MODULES) ? Strings.nullToEmpty((String) params.get(ValidationConfiguration.MODULES)) : "";
```

File: snomed/com.b2international.snowowl.validation.snomed/src/main/resources/scripts/rule482.groovy [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+import com.b2international.snowowl.snomed.datastore.request.SnomedRequests
+import com.google.common.base.Strings
+import com.google.common.collect.ImmutableMap
+
+import groovy.transform.Field
+
+List<ComponentIdentifier> issues = new ArrayList<>()
+
+def modules = SnomedRequests.prepareSearchConcept()
+	.filterByEcl(params.workingModules)
```
Comment (nagyo) [inline snomed/com.b2international.snowowl.validation.snomed/src/main/resources/scripts/rule482.groovy]: One more variable to rename

**Review by nagyo — CHANGES_REQUESTED** [review]
Please fix these two remaining items and we are good to go!

File: snomed/com.b2international.snowowl.validation.snomed/src/main/resources/scripts/rule482.groovy (hunk 8 above)
Comment (molnarlaura) [reply snomed/com.b2international.snowowl.validation.snomed/src/main/resources/scripts/rule482.groovy]: Done

**Review by nagyo — APPROVED** [review]

**Pull request merged**